### PR TITLE
Feature/metadata

### PR DESCRIPTION
**Pull request by jokedurnez**
Adding metadata information for monitoring.

Depends on https://github.com/openlattice/chronicle-server/pull/86, so leaving as draft for now.

image

### CODE REVIEW

**Review by anzioka — COMMENTED** [review]
looks good. how do you feel about centering daysCollected row content though?

Screen Shot 2020-06-18 at 7 02 30 PM

```
const CellContent = styled.div`
  display: flex;
  font-size: 15px;
  font-weight: 300;
  overflow: hidden;
  padding: 0 5px;
  justify-content: ${(props) => (props.centerContent ? 'center' : 'flex-start')}`;
```

Then 

```
<CellContent centerContent>
   { numDays }
</CellContent>
```

File: src/containers/study/utils/tableHeaders.js [hunk 1]
```diff
@@ -2,20 +2,37 @@
 
 import { PROPERTY_TYPE_FQNS } from '../../../core/edm/constants/FullyQualifiedNames';
 
-const { PERSON_ID, DATE_ENROLLED } = PROPERTY_TYPE_FQNS;
+const {
+  PERSON_ID, DATE_FIRST_PUSHED, DATE_LAST_PUSHED, EVENT_COUNT
+} = PROPERTY_TYPE_FQNS;
```
Comment (UnsungHero97) [inline src/containers/study/utils/tableHeaders.js]: Formatting:

```
const {
  DATE_FIRST_PUSHED, 
  DATE_LAST_PUSHED, 
  EVENT_COUNT,
  PERSON_ID, 
} = PROPERTY_TYPE_FQNS;
```

File: src/containers/study/utils/tableHeaders.js [hunk 2]
```diff
@@ -2,20 +2,37 @@
...
+    cellStyle: {
+      width: '22%'
+    }
+  },
+  {
+    key: EVENT_COUNT,
+    label: 'Days collected',
+    cellStyle: {
+      width: '18%',
+      textAligh: 'center',
```
Comment (UnsungHero97) [inline src/containers/study/utils/tableHeaders.js]: `textAlign`

File: src/containers/studies/StudiesSagas.js [hunk 3]
```diff
@@ -113,12 +114,17 @@ const {
...
 
-const { PARTICIPATED_IN, CHRONICLE_PARTOF } = ASSOCIATION_ENTITY_SET_NAMES;
+const { PARTICIPATED_IN, CHRONICLE_PARTOF, CHRONICLE_HAS } = ASSOCIATION_ENTITY_SET_NAMES;
 
 const {
   DATE_ENROLLED,
+  DATE_LOGGED,
+  DATE_LAST_PUSHED,
+  EVENT_COUNT,
+  // NOTIFICATION_DESCRIPTION,
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesSagas.js]: Please remove if unused.

File: src/containers/studies/StudiesReducer.js [hunk 4]
```diff
@@ -265,6 +270,20 @@ export default function studiesReducer(state :Map<*, *> = INITIAL_STATE, action
       });
     }
 
+    case getParticipantsMetadata.case(action.type): {
+      const seqAction :SequenceAction = action;
+      return getParticipantsMetadata.reducer(state, action, {
+        REQUEST: () => state.setIn([GET_PARTICIPANTS_METADATA, 'requestState'], RequestStates.PENDING),
+        FAILURE: () => state.setIn([GET_PARTICIPANTS_METADATA, 'requestState'], RequestStates.FAILURE),
+        SUCCESS: () => {
+          const { associationKeyIds, participantsEntitySetName } = seqAction.value;
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesReducer.js]: `associationKeyIds` is not passed on success...

```
getParticipantsMetadata.success(action.id, { neighborKeyIds, participantsEntitySetName })
```

File: src/containers/studies/StudiesSagas.js [hunk 5]
```diff
@@ -363,16 +438,36 @@ function* getStudyParticipantsWorker(action :SequenceAction) :Generator<*, *, *>
...
     );
     if (response.error) throw response.error;
     const enrollmentStatus :Map = response.data;
 
+    // get participant metadata
+    response = yield call(
+      getParticipantsMetadataWorker,
+      getParticipantsMetadata({ participants, participantsEntitySetId, participantsEntitySetName })
+    );
+    if (response.error) throw response.error;
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesSagas.js]: If the metadata request is the only request that fails, that means we can still show participants in the table, just without metadata. However, if you throw here, the the entire saga will fail, which means the participants table will be empty.

As such, I don't think you should throw here. I think it's better to show participants without metadata vs no participants at all.

File: src/containers/studies/StudiesSagas.js [hunk 6]
```diff
@@ -363,16 +438,36 @@ function* getStudyParticipantsWorker(action :SequenceAction) :Generator<*, *, *>
...
+    const metadata :Map = response.data;
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesSagas.js]: Assuming you update the code above and do not throw, then `response.data` will be undefined. You'll need to add a default here.

```
const metadata :Map = response.data || Map();
```

File: src/containers/studies/StudiesSagas.js [hunk 7]
```diff
@@ -363,16 +438,36 @@ function* getStudyParticipantsWorker(action :SequenceAction) :Generator<*, *, *>
...
+
     // update participants with enrollment status
-    participants = participants.map((participant, id) => participant
-      .set(STATUS, [enrollmentStatus.getIn([id, STATUS, 0], ENROLLED)])
-      .set(DATE_ENROLLED, [enrollmentStatus.getIn([id, DATE_ENROLLED, 0])])
-      .set('id', [id])); // required by LUK table
+    participants = participants.map((participant, id) => {
+
+      // If participant doesn't have metadata neighbor
+      const datesLogged = metadata.getIn([id, DATE_LOGGED]);
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesSagas.js]: Similarly, assuming `metadata` is empty, you should pass a default value for this `.getIn()`.

```
const datesLogged :List = metadata.getIn([id, DATE_LOGGED], List());
const count :number = datesLogged.count();
```

File: src/containers/studies/StudiesSagas.js [hunk 8]
```diff
@@ -363,16 +438,36 @@ function* getStudyParticipantsWorker(action :SequenceAction) :Generator<*, *, *>
...
+    );
+    if (response.error) throw response.error;
+    const metadata :Map = response.data;
+
     // update participants with enrollment status
-    participants = participants.map((participant, id) => participant
-      .set(STATUS, [enrollmentStatus.getIn([id, STATUS, 0], ENROLLED)])
-      .set(DATE_ENROLLED, [enrollmentStatus.getIn([id, DATE_ENROLLED, 0])])
-      .set('id', [id])); // required by LUK table
+    participants = participants.map((participant, id) => {
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesSagas.js]: Please rename `id` to `participantEKID` or `participantEntityKeyId`.

File: src/containers/studies/StudiesSagas.js [hunk 9]
```diff
@@ -363,16 +438,36 @@ function* getStudyParticipantsWorker(action :SequenceAction) :Generator<*, *, *>
...
+        count = datesLogged.size;
+      }
+
+      return participant
+        .set(STATUS, [enrollmentStatus.getIn([id, STATUS, 0], ENROLLED)])
+        .set(DATE_ENROLLED, [enrollmentStatus.getIn([id, DATE_ENROLLED, 0])])
+        .set(DATE_LAST_PUSHED, [metadata.getIn([id, DATE_LAST_PUSHED, 0])])
+        .set(EVENT_COUNT, [count])
+        .set('id', [id]);
+    }); // required by LUK table
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesSagas.js]: Please move the `// required by LUK table` comment to be after the `.set('id', [id]);`, which is the actual LUK requirement.

**Review by UnsungHero97 — CHANGES_REQUESTED** [review]

File: src/containers/studies/StudiesSagas.js (hunk 7 above)
Comment (jokedurnez) [reply src/containers/studies/StudiesSagas.js]: Done.  Out of curiosity, why `count()` and not `.size`?

File: src/containers/studies/StudiesActions.js [hunk 10]
```diff
@@ -29,6 +29,9 @@ const getGlobalNotificationsEKID = newRequestSequence(GET_GLOBAL_NOTIFICATIONS_E
 const GET_PARTICIPANTS_ENROLLMENT :'GET_PARTICIPANTS_ENROLLMENT' = 'GET_PARTICIPANTS_ENROLLMENT';
 const getParticipantsEnrollmentStatus :RequestSequence = newRequestSequence(GET_PARTICIPANTS_ENROLLMENT);
 
+const GET_PARTICIPANTS_METADATA :'GET PARTICIPANTS_METADATA' = 'GET_PARTICIPANTS_METADATA';
+const getParticipantsMetadata :RequestSequence = newRequestSequence(GET_PARTICIPANTS_METADATA);
+
```
Comment (anzioka) [inline src/containers/studies/StudiesActions.js]: `GET_PARTICIPANTS_METADATA`

File: src/containers/studies/StudiesSagas.js [hunk 11]
```diff
@@ -106,12 +107,19 @@ const { PermissionTypes, UpdateTypes } = Types;
...
+  CHRONICLE_STUDIES,
+  CHRONICLE_METADATA,
+} = ENTITY_SET_NAMES;
 
-const { PARTICIPATED_IN, PART_OF_ES_NAME } = ASSOCIATION_ENTITY_SET_NAMES;
+const { PARTICIPATED_IN, PART_OF_ES_NAME, HAS_ES_NAME } = ASSOCIATION_ENTITY_SET_NAMES;
 
 const {
   DATE_ENROLLED,
+  DATE_LOGGED,
```
Comment (anzioka) [inline src/containers/studies/StudiesSagas.js:120]: missing from PROPERTY_TYPE_FQNS object

File: src/containers/study/components/ParticipantRow.js [hunk 12]
```diff
@@ -48,13 +51,14 @@ const RowWrapper = styled.tr.attrs(() => ({ tabIndex: '1' }))`
...
 /* stylelint-disable value-no-vendor-prefix, property-no-vendor-prefix */
 const CellContent = styled.div`
-  -webkit-line-clamp: 2;
-  display: -webkit-box;
+  display: flex;
   font-size: 15px;
   font-weight: 300;
   overflow: hidden;
   padding: 0 5px;
+  justify-content: ${(props) => (props.centerContent ? 'center' : 'flex-start')}
```
Comment (anzioka) [inline src/containers/study/components/ParticipantRow.js]: missing `;` at the end of this line

**Review by anzioka — CHANGES_REQUESTED** [review]

File: src/containers/studies/StudiesSagas.js (hunk 7 above)
Comment (UnsungHero97) [reply src/containers/studies/StudiesSagas.js]: https://github.com/immutable-js/immutable-js/pull/1560

File: src/containers/studies/StudiesReducer.js [hunk 13]
```diff
@@ -265,6 +270,20 @@ export default function studiesReducer(state :Map<*, *> = INITIAL_STATE, action
...
+        REQUEST: () => state.setIn([GET_PARTICIPANTS_METADATA, 'requestState'], RequestStates.PENDING),
+        FAILURE: () => state.setIn([GET_PARTICIPANTS_METADATA, 'requestState'], RequestStates.FAILURE),
+        SUCCESS: () => {
+          const { neighborKeyIds, participantsEntitySetName } = seqAction.value;
+          return state
+            .setIn(['neighborKeyIds', participantsEntitySetName], neighborKeyIds)
+            .setIn([GET_PARTICIPANTS_METADATA, 'requestState'], RequestStates.SUCCESS);
+        }
+      });
+    }
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesReducer.js]: I don't see `neighborKeyIds` or the `GET_PARTICIPANTS_METADATA ` request state being used in any component, so I'm pretty sure all of the changes to this file can be removed.

**Review by UnsungHero97 — CHANGES_REQUESTED** [review]

**Review by UnsungHero97 — APPROVED** [review]

**Review by anzioka — APPROVED** [review]

**Pull request merged**